packages/react-dom/src/client/DOMPropertyOperations.js

```diff
@@ -149,9 +151,52 @@ export function setValueForProperty(
   if (shouldIgnoreAttribute(name, propertyInfo, isCustomComponentTag)) {
     return;
   }
+
+  if (
+    enableCustomElementPropertySupport &&
+    isCustomComponentTag &&
+    name[0] === 'o' &&
+    name[1] === 'n'
+  ) {
+    let eventName = name.replace(/Capture$/, '');
+    const useCapture = name !== eventName;
+    const nameLower = eventName.toLowerCase();
+    if (nameLower in node) {
+      eventName = nameLower;
+    }
+    eventName = eventName.slice(2);
+
+    const listenersObjName = eventName + (useCapture ? 'true' : 'false');
+    const listeners = getCustomElementEventHandlersFromNode(node);
+    const alreadyHadListener = listeners[listenersObjName];
+
+    if (typeof value === 'function' || alreadyHadListener) {
+      listeners[listenersObjName] = value;
+      const proxy = useCapture ? fireEventProxyCapture : fireEventProxy;
+      if (value) {
+        if (!alreadyHadListener) {
+          node.addEventListener(eventName, proxy, useCapture);
+        }
+      } else {
+        node.removeEventListener(eventName, proxy, useCapture);
+      }
+      return;
+    }
+  }
+
   if (shouldRemoveAttribute(name, value, propertyInfo, isCustomComponentTag)) {
     value = null;
   }
+
+  if (
+    enableCustomElementPropertySupport &&
+    isCustomComponentTag &&
+    name in (node: any)
+  ) {
+    (node: any)[name] = value;
```

Comment:
We should have test coverage verifying that this _does not_ get set in cases where we don't expect. E.g. if `oncustomevent` is _not_ supposed to set the `oncustomevent` property even when it exists, we should have a test verifying this. That test should specify what happens when the value type changes midway (e.g. function -> string and back).

Reply:
Thanks for pointing this out!
I added some tests with a variety of switching between nothing, string, and function.